worker/peergrouper/worker.go

```diff
@@ -350,26 +357,46 @@ func inStrings(t string, ss []string) bool {
 	return false
 }
 
-func (w *pgWorker) apiPublishInfo() ([][]network.HostPort, error) {
-	details := apiserver.Details{
-		Servers:   make(map[string]apiserver.APIServer),
-		LocalOnly: true,
-	}
-	servers := make([][]network.HostPort, 0, len(w.machineTrackers))
+// apiserverHostPorts returns the host-ports for each apiserver machine.
+func (w *pgWorker) apiserverHostPorts() map[string][]network.HostPort {
+	servers := make(map[string][]network.HostPort)
 	for _, m := range w.machineTrackers {
 		hostPorts := network.AddressesWithPort(m.Addresses(), w.config.APIPort)
-		server := apiserver.APIServer{ID: m.Id()}
 		if len(hostPorts) == 0 {
 			continue
 		}
+		servers[m.Id()] = hostPorts
+	}
+	return servers
+}
+
+func (w *pgWorker) publishAPIServerDetails(
+	servers map[string][]network.HostPort,
+	members map[string]replicaset.Member,
+) {
+	details := apiserver.Details{
+		Servers:   make(map[string]apiserver.APIServer),
+		LocalOnly: true,
+	}
+	for id, hostPorts := range servers {
+		var internalAddress string
+		mongoAddress, _, err := net.SplitHostPort(members[id].Address)
```

Comment:
This is a consequence of the peergrouper being co-opted. Its primary duty, to date, has been to maintain the mongo replicaset configuration. I think it'd be a little awkward to to track it explicitly, without more invasive changes, because this address is taken from the replicaset.Member structure.